File: Swift/프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.swift
import Foundation

func solution(_ arr:[[Int]]) -> [Int] {
    let n = arr.count
    let m = arr[0].count
    
    var num_0 = 0
    var num_1 = 0
    var minus = 0
    
    for i in 0..<n {
        for j in 0..<m {
            if arr[i][j] == 0 {
                num_0 += 1
            } else {
                num_1 += 1
            }
        }
    }
    
    func check(_ i: Int, _ j: Int, _ size: Int) -> Bool {
        var cnt = 0
        if i >= n && j >= m {
            return false
        }
        var k = arr[i][j]
        for dx in i..<i+size {
            for dy in j..<j+size {
                cnt += 1
                if arr[dx][dy] != k {
                    return false
                }
            }
        }
        
        if k == 1 {
            num_1 -= (cnt - 1)
        } else {
            num_0 -= (cnt - 1)
        }
        return true
    }
    
    func quad(_ i: Int, _ j: Int, _ size: Int) {
        if size == 1 {
            return
        }
        
        if check(i, j, size) {
            return
        }
        
        let newSize = size / 2
        quad(i, j, newSize)           // Top-left
        quad(i, j + newSize, newSize) // Top-right
        quad(i + newSize, j, newSize) // Bottom-left
        quad(i + newSize, j + newSize, newSize) // Bottom-right
    }
    
    quad(0,0,n)
    return [num_0, num_1]
}

File: Python3/프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.py
def solution(arr):
    # 초기 0,1 개수 셋팅
    cnt_0 = 0
    cnt_1 = 0
    for i in arr:
        for j in i:
            if j == 0:
                cnt_0 += 1
            else:
                cnt_1 += 1
    
    def check(x,y,size):
        # print("->   check")
        flag = None
        cnt = 0
        for i in range(x, x+size):
            for j in range(y, y+size):
                # print("                                     ",arr[i][j])
                if flag == None:
                    flag = arr[i][j]
                else:
                    cnt += 1
                    if flag != arr[i][j]:
                        return False
        
        if flag == 1:
            nonlocal cnt_1
            cnt_1 -= cnt
        else:
            nonlocal cnt_0
            cnt_0 -= cnt

        # print("->   cnt", cnt)
        return True
    
    def backtracking(x, y, size):
        # print(x,y,size)
        if size == 1:
            return
        if check(x,y,size):
            return
        
        half = size // 2
        backtracking(x, y, half) # 1사분면
        backtracking(x + half, y, size // 2) # 2사분면
        backtracking(x, y + half, size // 2) # 3사분면
        backtracking(x + half, y + half, size // 2) # 4사분면
    
    backtracking(0, 0, len(arr))
    return [cnt_0, cnt_1]